src/mol-theme/color/external-structure.ts

```diff
@@ -0,0 +1,157 @@
+
+function getStyleTheme(ctx: ThemeDataContext, props: ExternalStructureColorThemeProps['style']) {
+    switch (props.name) {
+        case 'chain-id': return ChainIdColorTheme(ctx, props.params);
+        case 'entity-id': return EntityIdColorTheme(ctx, props.params);
+        case 'entity-source': return EntitySourceColorTheme(ctx, props.params);
+        case 'molecule-type': return MoleculeTypeColorTheme(ctx, props.params);
+        case 'model-index': return ModelIndexColorTheme(ctx, props.params);
+        case 'structure-index': return StructureIndexColorTheme(ctx, props.params);
+        default: assertUnreachable(props);
+    }
+}
+
+export function ExternalStructureColorTheme(ctx: ThemeDataContext, props: PD.Values<ExternalStructureColorThemeParams>): ColorTheme<ExternalStructureColorThemeParams> {
+    let structure: Structure | undefined;
+    try {
+        structure = props.structure.getValue();
+    } catch {
+        // .getValue() is resolved during state reconciliation => would throw from UI
+    }
+
+    // NOTE: this will currently be slow for with GPU/texture meshes due to slow iteration
+    // TODO: create texture to be able to do the sampling on the GPU
+
+    let color: LocationColor;
+    let contextHash: number | undefined = undefined;
+    let legend: Readonly<ScaleLegend | TableLegend> | undefined = undefined;
+
+    const { maxDistance, takeFirstDistance, normalOffset, defaultColor, backboneOnly } = props;
+
+    if (structure) {
+        const styleTheme = getStyleTheme({ ...ctx, structure }, props.style);
+
+        const lookupFirstCtx = StructureLookup3DResultContext();
+        const lookupNearestCtx = StructureLookup3DResultContext();
+
+        let s = structure;
+        if (backboneOnly) {
+            s = StructureSelection.unionStructure(StructureSelectionQueries.backbone.query(new QueryContext(structure)));
+        }
+
+        const position = Vec3();
+        const l = StructureElement.Location.create(s);
+
+        color = (location: Location, isSecondary: boolean): Color => {
+            if (!isPositionLocation(location)) {
+                return defaultColor;
+            }
+
+            // Offset the vertex position along its normal
+            if (normalOffset !== 0) {
+                Vec3.scaleAndAdd(position, location.position, location.normal, normalOffset);
+            } else {
+                Vec3.copy(position, location.position);
+            }
+
+            const [x, y, z] = position;
+            if (!s.lookup3d.check(x, y, z, maxDistance)) return defaultColor;
+
+            const rf = s.lookup3d.first(x, y, z, takeFirstDistance, lookupFirstCtx);
+            if (rf.count > 0) {
+                l.unit = rf.units[0];
+                l.element = l.unit.elements[rf.indices[0]];
+                return styleTheme.color(l, isSecondary);
+            }
+
+            const rn = s.lookup3d.nearest(x, y, z, 1, lookupNearestCtx);
+            if (rn.count === 0) return defaultColor;
```

Comment:
I can imagine this could get rather slow. Would it make sense to add a grid mode for this? This would trade memory for linear lookup complexity by preparing a coloring grid ahead of time.

Reply:
It is quite OK. For my example, PDB entry 9C5B, it takes 200ms with default parameters on my machine. Filling a 3D grid at comparable resolution will take longer. A grid mode could be beneficial if the isolevel needs to be changed often and smoothly.